server/cluster/cluster.go

```diff
@@ -630,3 +649,84 @@ func (c *Cluster) GetNodes(_ context.Context) (*GetNodesResult, error) {
 		NodeShards:             nodeShards,
 	}, nil
 }
+
+func (c *Cluster) GetClusterVersion() uint64 {
+	c.lock.RLock()
+	defer c.lock.RUnlock()
+	return c.metaData.clusterTopology.Version
+}
+
+func (c *Cluster) GetClusterMinNodeCount() uint32 {
+	c.lock.RLock()
+	defer c.lock.RUnlock()
+	return c.metaData.cluster.MinNodeCount
+}
+
+func (c *Cluster) GetClusterShardTotal() uint32 {
+	c.lock.RLock()
+	defer c.lock.RUnlock()
+	return c.metaData.cluster.ShardTotal
+}
+
+func (c *Cluster) GetClusterState() clusterpb.ClusterTopology_ClusterState {
+	c.lock.RLock()
+	defer c.lock.RUnlock()
+	return c.metaData.clusterTopology.State
+}
+
+func (c *Cluster) GetClusterShardView() []*clusterpb.Shard {
+	c.lock.RLock()
+	defer c.lock.RUnlock()
+	return c.metaData.clusterTopology.ShardView
+}
+
+func (c *Cluster) UpdateClusterTopology(ctx context.Context, state clusterpb.ClusterTopology_ClusterState, shardView []*clusterpb.Shard) error {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+	c.metaData.clusterTopology.ShardView = shardView
+	c.metaData.clusterTopology.State = state
+	return c.GetStorage().PutClusterTopology(ctx, c.GetClusterID(), c.GetClusterVersion(), c.metaData.clusterTopology)
+}
+
+type ShardWithLock struct {
+	shardID uint32
+	lock    sync.Mutex
+}
+
+func (c *Cluster) getShardLock(shardID uint32) *ShardWithLock {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+	_, ok := c.shardLockMap[shardID]
+	if !ok {
+		c.shardLockMap[shardID] = &ShardWithLock{shardID: shardID, lock: sync.Mutex{}}
+	}
+	return c.shardLockMap[shardID]
+}
+
+func (c *Cluster) LockShardByID(shardID uint32) bool {
+	shardLock := c.getShardLock(shardID)
+	return shardLock.lock.TryLock()
+}
+
+func (c *Cluster) LockShardByIDWithRetry(shardID uint32, maxRetrySize int, waitDuration time.Duration) bool {
```

Comment:
There is no need to suffix the variable name with the type:
```suggestion
func (c *Cluster) LockShardByIDWithRetry(shardID uint32, maxRetry int, waitInterval time.Duration) bool {
```